### ISSUE TITLE
http-client should allow to log requests and responses

### ISSUE THREAD

**Issue by loewenstein** [snoyberg/http-client#248]
I would like to be able to see all requests and responses in a debug log (e.g. on STDERR or STDOUT).

I have seen https://github.com/snoyberg/http-client/blob/master/http-client/Network/HTTP/Client/Types.hs#L672 and https://github.com/snoyberg/http-client/blob/master/http-client/Network/HTTP/Client/Types.hs#L497. However running the BodyReader and printing the result consumes the body, i.e. it is no longer available for the issuer of the request.

If the two hooks are actually the right ones to use (I doubt that, it looked like a workaround in the beginning), we should be able to modify the Response (just as we can modify the request). That way one could feed the body back into a new otherwise identical Response.
This would have to be solved even if we provide separate hooks though.

What do you think?

With some amount of guidance I might be able to come up with a PR.

Edit: Maybe I should add, I am using http-client through servant-client. Hence my ability to just call a different request execution method is rather limited.

**Comment by snoyberg** [comment]
These definitely weren't intended for the logging use case, but if it fits, I'd probably say to stick with that. If you really need to log the response bodies, I'd be open to an extra method in ManagerSettings for post-processing a `Response` value.

**Comment by loewenstein** [comment]
Well, it doesn't work because the body gets consumed from me printing it to stderr. At least that is my interpretation of what I see.

"post-processing a `Response` value" would be a separate method like `managerModifyRequest`, i.e. `managerModifyResponse :: ManagerSettings -> Response -> IO Response`?

So logging would go into the two separate modify methods. Sounds good to me.
If you have a pointer to where these get called should get called I would be happy to propose a PR.

I guess the logging functionality would have to consume the body and provide a Response with an in memory version of the body, right? I guess I should limit body logging to a limited `Content-Length` then.

**Comment by snoyberg** [comment]
> Well, it doesn't work because the body gets consumed from me printing it to stderr. At least that is my interpretation of what I see.

I agree. That's what I was getting at with "if you need to log the response bodies." Many use cases would simply want to log the response metadata (status/headers), in which case the current API is sufficient. But your use case may not be that.

Yes, I think `managerModifyResponse` is the right thing. And it sounds like a PR would only need to add that one method, and it could be called from the same place as `checkResponse`.

You're correct about the in memory representation, though an alternative would be streaming to a temporary file. But to be clear, I think the logging functionality itself would _not_ end up in http-client, correct?

**Comment by loewenstein** [comment]
I immediately get the following error about missing header files.

Is there a contributor guide that would tell me how to prepare my machine to build and test http-client?

```
12:35 $ stack build
HsOpenSSL-0.11.1.1: configure
HsOpenSSL-0.11.1.1: build
Progress: 1/2
--  While building package HsOpenSSL-0.11.1.1 using:
      /Users/d043856/.stack/setup-exe-cache/x86_64-osx/setup-Simple-Cabal-1.22.5.0-ghc-7.10.3 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.22.5.0 build --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/d043856/projects/http-client/.stack-work/logs/HsOpenSSL-0.11.1.1.log

    Configuring HsOpenSSL-0.11.1.1...
    Building HsOpenSSL-0.11.1.1...
    Preprocessing library HsOpenSSL-0.11.1.1...
    In file included from BN.hsc:45:
    /private/var/folders/88/__cnr5f92cbd7zshg3kvp_yj1199kq/T/stack81527/HsOpenSSL-0.11.1.1/cbits/HsOpenSSL.h:3:10: fatal error: 'openssl/asn1.h' file not found
    #include <openssl/asn1.h>
             ^
    1 error generated.
    compiling .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build/OpenSSL/BN_hsc_make.c failed (exit code 1)
    command was: /usr/bin/gcc -c .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build/OpenSSL/BN_hsc_make.c -o .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build/OpenSSL/BN_hsc_make.o -m64 -fno-stack-protector -D__GLASGOW_HASKELL__=710 -Ddarwin_BUILD_OS=1 -Dx86_64_BUILD_ARCH=1 -Ddarwin_HOST_OS=1 -Dx86_64_HOST_ARCH=1 -Icbits -D PTHREAD -DCALLCONV=ccall -I.stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build/autogen -include .stack-work/dist/x86_64-osx/Cabal-1.22.5.0/build/autogen/cabal_macros.h -I/Users/d043856/.stack/snapshots/x86_64-osx/lts-5.10/7.10.3/lib/x86_64-osx-ghc-7.10.3/network-2.6.2.1-A6pE0dLgMEk2GtUa1JtTJQ/include -I/Users/d043856/.stack/programs/x86_64-osx/ghc-7.10.3/lib/ghc-7.10.3/unix_KZL8h98IqDM57kQSPo1mKx/include -I/Users/d043856/.stack/programs/x86_64-osx/ghc-7.10.3/lib/ghc-7.10.3/time_FTheb6LSxyX1UABIbBXRfn/include -I/Users/d043856/.stack/programs/x86_64-osx/ghc-7.10.3/lib/ghc-7.10.3/bytes_6VWy06pWzJq9evDvK2d4w6/include -I/Users/d043856/.stack/programs/x86_64-osx/ghc-7.10.3/lib/ghc-7.10.3/base_HQfYBxpPvuw8OunzQu6JGM/include -I/Users/d043856/.stack/programs/x86_64-osx/ghc-7.10.3/lib/ghc-7.10.3/integ_2aU3IZNMF9a7mQ0OzsZ0dS/include -I/Users/d043856/.stack/programs/x86_64-osx/ghc-7.10.3/lib/ghc-7.10.3/include -I/Users/d043856/.stack/programs/x86_64-osx/ghc-7.10.3/lib/ghc-7.10.3/include/
```

**Comment by snoyberg** [comment]
Just try stack test http-client

**Comment by loewenstein** [comment]
Worked. Thanks.

Now looking into the code there is one more thing to consider.
`Response body` is a parameterized type. What should `body` be in `managerModifyResponse :: ManagerSettings -> Response body -> IO Response body`.

I guess `BodyReader` is the right thing, because that's what `checkResponse` has as well. However I don't have a holistic view on the project and would like to know (at least a little) why this is the right thing.

**Comment by snoyberg** [comment]
Because the internal engine always generates a `Response BodyReader`, and it's only upper layers (like `http-conduit`) that convert it to a different representation.

**Comment by snoyberg** [comment]
Implemented by PR #249, thanks!
